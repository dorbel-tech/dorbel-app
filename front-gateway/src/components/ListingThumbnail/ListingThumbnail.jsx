import React, { Component } from 'react';
import { inject, observer } from 'mobx-react';
import './ListingThumbnail.scss';
import { Col } from 'react-bootstrap';
import NavLink from '~/components/NavLink';
import ListingBadge from '../ListingBadge/ListingBadge';
import CloudinaryImage from '../CloudinaryImage/CloudinaryImage';
import LikeButton from '../LikeButton/LikeButton';
import utils from '../../providers/utils';

@inject('appStore') @observer
class ListingThumbnail extends Component {
  getListingUrl(listing) {
    if (this.props.isMyProperties) {
      return utils.getDashPropertyPath(listing);
    } else {
      return utils.getListingPath(listing);
    }    
  }

  getOheLabel() {
    const { listing, appStore }  = this.props;

    if (appStore.oheStore.isListingLoaded(listing.id)) {
      const oheCount = appStore.oheStore.oheByListingId(listing.id).filter(ohe => ohe.status === 'open').length;
      if (oheCount) {
        return <span className="pull-left apt-thumb-ohe-text">{oheCount} מועדי ביקור זמינים</span>;
      } else {
        return (
          <span className="pull-left">
            <span className="apt-thumb-no-ohe">אין מועדי ביקור</span>
            &nbsp;
            <span className="apt-thumb-ohe-text">עדכנו אותי</span>
          </span>
        );
      }
    } else {
      return null;
    }
  }

  render() {
    const { listing } = this.props;
    const sortedListingImages = utils.sortListingImages(listing);
    const imageURL = sortedListingImages.length ? sortedListingImages[0].url : '';
    const classLeaseDate = new Date(listing.lease_start) <= Date.now() ? 'apt-thumb-lease-immediate' : 'apt-thumb-lease-date';
    const listingDateStr = new Date(listing.lease_start) <= Date.now() ? 'מיידי' : utils.formatDate(listing.lease_start);

    return (
      <Col lg={4} sm={6} xs={12}>
        <NavLink to={this.getListingUrl(listing)}
          className="thumbnail apt-thumb-container apt-thumb-container-single pull-right">
          <ListingBadge listing={listing} />
          <div className="apt-thumb-apt-image">
            <CloudinaryImage src={imageURL} height={500} />
          </div>
          <div className="apt-thumb-details">
            <div className="apt-thumb-details-header">
              <div className="apt-thumb-details-title">
                {utils.getListingTitle(listing)}
              </div>
              <div className="apt-thumb-details-like">
                <LikeButton listingId={listing.id} />
              </div>
            </div>
            <div className="apt-thumb-details-address">
              {utils.getListingSubTitle(listing)}
            </div>
            <div className="apt-thumb-details-extra">
              <span>
                {listing.apartment.size}</span><span className="apt-thumb-sub-text"> מ״ר</span>
              <span className="apt-thumb-details-extra-rooms">
                {listing.apartment.rooms}</span><span className="apt-thumb-sub-text"> חד'</span>
              <span className="apt-thumb-details-extra-date">
                <span className="apt-thumb-sub-text">תאריך כניסה </span>
                <span className={classLeaseDate}>{listingDateStr}</span>
              </span>
            </div>
          </div>
          <div className="apt-thumb-caption">
            {listing.monthly_rent}<span className="apt-thumb-sub-text"> ₪</span>
            { this.getOheLabel() }
          </div>
        </NavLink>
      </Col>
    );
  }
}

ListingThumbnail.wrappedComponent.propTypes = {
  listing: React.PropTypes.object.isRequired,
  appStore: React.PropTypes.object.isRequired,
  isMyProperties: React.PropTypes.string
};

export default ListingThumbnail;
