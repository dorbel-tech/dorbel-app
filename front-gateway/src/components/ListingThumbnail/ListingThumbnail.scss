@import '../sharedStyles/variables.scss';
@import "../sharedStyles/bsvars.scss";

.apt-thumb-container {
  padding: 40px 0;
  background-color: $brand-white;
  border-radius: 4px 4px 0 0;
  box-shadow: 0 2px 15px 0 rgba(0, 0, 0, 0.1);
  opacity: 1 !important;

  h5 {
    font-weight: 600;
    letter-spacing: 0.7px;
    padding: 5px;
  }

  .related-listings-title {
    margin: 0 0 25px;
  }
}

.apt-thumb-container-single {
  padding: 0;
  text-align: right;
  width: 100%;

  .listingBadge {
    border-radius: 100px 0 0 100px;
    min-width: 84px;
    position: absolute;
    top: 150px;
  }

  @media (min-width: $screen-sm-min) {
    max-width: 292px;
  }

  @media (min-width: $screen-md-min) {
    transition: box-shadow 0.35s ease-out, transform 0.3s ease-out, opacity 0.2s ease-out;
  }

  &:hover {
    text-decoration: none;

    @media (min-width: $screen-md-min) {
      box-shadow: 0 15px 20px rgba(0, 0, 0, 0.2);
      transform: translate(0, -4px);
    }

    .apt-thumb-apt-image {
      object-fit: cover;
    }
  }

  .apt-thumb-caption {
    color: $gray-dark;
    font-size: 18px;
    line-height: 18px;
    padding: 15px;

    .apt-thumb-ohe-text {
      color: #2f93dc;
      font-size: 13px;
      text-decoration: underline;

      &:hover {
        color: #006090;
      }
    }

    .apt-thumb-no-ohe {
      color: #9b9b9b;
      font-size: 13px;
    }
  }
}

.apt-thumb-apt-image img {
  @include border-top-radius(3px);

  height: 195px;
  object-fit: cover;
  overflow: hidden;
  width: 100%;
}

.apt-thumb-sub-text {
  color: #9b9b9b;
  font-size: 11px;
}

.apt-thumb-warning-sub-text {
  color: #ea441a;
}

.apt-thumb-lease-immediate {
  color: $brand-green;
}

.apt-thumb-details {
  border-bottom: $gray-border;
  padding: 15px;

  .apt-thumb-details-header {
    .apt-thumb-details-title {
      color: #4d4d4d;
      font-size: 18px;
      height: 23px;
      line-height: 21px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      display: inline-block;
      width: 90%;
    }
  }

  .apt-thumb-details-address {
    color: #4a4a4a;
    font-size: 13px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .apt-thumb-details-extra {
    color: #4d4d4d;
    font-size: 13px;
    padding-top: 15px;

    .apt-thumb-details-extra-rooms {
      padding-right: 13px;
    }

    .apt-thumb-details-extra-date {
      color: $gray-dark;
      float: left;
    }
  }
}

// TODO: temporary class for AB testing
.apt-thumb-details-header .apt-thumb-details-like {
  display: none;
}

// TODO: temporary class for AB testing
.apt-thumb-caption .apt-thumb-details-like {
  display: block;
}

.apt-thumb-details-like {
  float: left;

  .like-button.like-button-with-text,
  .like-button.like-button-with-text.like-button-liked {
    background-color: $brand-white;
    border: none;
    color: #ea441a;
    height: auto;
    width: auto;

    .like-button-text {
      padding-left: 0;
    }

    i {
      color: #ea441a;
      padding: 0 8px;
    }
  }
}
