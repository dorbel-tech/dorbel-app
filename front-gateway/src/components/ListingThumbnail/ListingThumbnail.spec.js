'use strict';
import React from 'react';
import { shallow } from 'enzyme';

import ListingThumbnail from './ListingThumbnail';

describe('Listing Thumbnail', () => {
  let appStoreMock;

  beforeAll(() => {
    appStoreMock = {
      oheStore: {
        isListingLoaded: jest.fn(),
        oheByListingId: jest.fn().mockReturnValue([])
      },
      authStore: {
        isLoggedIn: () => false
      }
    };
  });

  describe('OHE indication', () => {
    function renderThumbnail() {
      const listing = {
        id: 777,
        images: [],
        apartment: {
          building: {
            neighborhood: {},
            city: {}
          }
        }
      };

      const rendered = shallow(<ListingThumbnail.wrappedComponent appStore={appStoreMock} listing={listing} />);
      return rendered.children().children(); // ListingThumbnail contains a Col which contains a NavLink
    }

    it('should display OHEs when they are already loaded', () => {
      const oheCount = 7;
      appStoreMock.oheStore.isListingLoaded.mockReturnValue(true);
      appStoreMock.oheStore.oheByListingId.mockReturnValue(Array(oheCount).fill({ status: 'open' }));

      const rendered = renderThumbnail();

      const oheIndication = rendered.find('.apt-thumb-ohe-text');
      expect(oheIndication.exists()).toBe(true);
      expect(oheIndication.text()).toBe(`${oheCount} מועדי ביקור זמינים`);
      expect(oheIndication).toMatchSnapshot();
    });

    it('should only display indication for open or registered OHEs', () => {
      const openCount = 3;
      const registeredCount = 2;
      const closedCount = 4;
      appStoreMock.oheStore.isListingLoaded.mockReturnValue(true);
      appStoreMock.oheStore.oheByListingId.mockReturnValue([].concat(
        Array(openCount).fill({ status: 'open' }),
        Array(closedCount).fill({ status: 'not open' }),
        Array(registeredCount).fill({ status: 'registered' })
      ));

      const rendered = renderThumbnail();

      const oheIndication = rendered.find('.apt-thumb-ohe-text');
      expect(oheIndication.exists()).toBe(true);
      expect(oheIndication.text()).toBe(`${openCount + registeredCount} מועדי ביקור זמינים`);
      expect(oheIndication).toMatchSnapshot();
    });

    it('should display indication for a loaded listing with no events', () => {
      appStoreMock.oheStore.isListingLoaded.mockReturnValue(true);
      appStoreMock.oheStore.oheByListingId.mockReturnValue([]);

      const rendered = renderThumbnail();

      const oheIndication = rendered.find('.apt-thumb-no-ohe');
      const oheLink = rendered.find('.apt-thumb-ohe-text');
      expect(oheIndication.exists()).toBe(true);
      expect(oheIndication.text()).toBe('אין מועדי ביקור');
      expect(oheIndication).toMatchSnapshot();
      expect(oheLink.exists()).toBe(true);
      expect(oheLink.text()).toBe('עדכנו אותי');
      expect(oheLink).toMatchSnapshot();
    });

    it('should not render anything when listing events are not loaded', () => {
      appStoreMock.oheStore.isListingLoaded.mockReturnValue(false);

      const rendered = renderThumbnail();

      const noOheIndication = rendered.find('.apt-thumb-no-ohe');
      const oheIndication = rendered.find('.apt-thumb-ohe-text');
      expect(noOheIndication.exists()).toBe(false);
      expect(oheIndication.exists()).toBe(false);
    });
  });
});
