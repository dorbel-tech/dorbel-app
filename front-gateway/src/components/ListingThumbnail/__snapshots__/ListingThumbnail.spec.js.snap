exports[`Listing Thumbnail OHE indication should display OHEs when they are already loaded 1`] = `
<span
  className="pull-left apt-thumb-ohe-text">
  7
   מועדי ביקור זמינים
</span>
`;

exports[`Listing Thumbnail OHE indication should display indication for a loaded listing with no events 1`] = `
<span
  className="pull-left apt-thumb-no-ohe">
  אין מועדי ביקור
</span>
`;

exports[`Listing Thumbnail OHE indication should only display indication for open or registered OHEs 1`] = `
<span
  className="pull-left apt-thumb-ohe-text">
  5
   מועדי ביקור זמינים
</span>
`;
